### PR TITLE
Add Ingress TLS and opaque secret support

### PR DESCRIPTION
**Pull request by theodoracheng**
Support opaque and tls ingress secrets https://cloud.ibm.com/docs/containers?topic=containers-secrets. This allows users to creates kubernetes TLS secrets with a cert crn backed by IBM Cloud Secrets Manager as well as create kubernetes opaque secrets with multiple fields that correlated to secrets in IBM Cloud Secrets Manager.

Also fixes https://github.com/IBM-Cloud/terraform-provider-ibm/issues/4552

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Output from acceptance testing:

```
make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainerIngressSecretTLS_Basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainerIngressSecretTLS_Basic -timeout 700m
=== RUN   TestAccIBMContainerIngressSecretTLS_Basic
--- PASS: TestAccIBMContainerIngressSecretTLS_Basic (39.97s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes	44.115s
```

```
make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainerIngressSecretTLSDatasourceBasic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainerIngressSecretTLSDatasourceBasic -timeout 700m
=== RUN   TestAccIBMContainerIngressSecretTLSDatasourceBasic
--- PASS: TestAccIBMContainerIngressSecretTLSDatasourceBasic (23.06s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes	27.366s
```

```
make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainerIngressSecretOpaque_Basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainerIngressSecretOpaque_Basic -timeout 700m
=== RUN   TestAccIBMContainerIngressSecretOpaque_Basic
--- PASS: TestAccIBMContainerIngressSecretOpaque_Basic (22.58s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes	26.302s
```

```
make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainerIngressSecretOpaqueDatasourceBasic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainerIngressSecretOpaqueDatasourceBasic -timeout 700m
=== RUN   TestAccIBMContainerIngressSecretOpaqueDatasourceBasic
--- PASS: TestAccIBMContainerIngressSecretOpaqueDatasourceBasic (14.30s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes	18.599s

```

```
make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainerIngressInstanceDatasourceBasic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainerIngressInstanceDatasourceBasic -timeout 700m
=== RUN   TestAccIBMContainerIngressInstanceDatasourceBasic
--- PASS: TestAccIBMContainerIngressInstanceDatasourceBasic (10.52s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes	14.208s
```

```
make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainerIngressInstance_Basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainerIngressInstance_Basic -timeout 700m
=== RUN   TestAccIBMContainerIngressInstance_Basic
--- PASS: TestAccIBMContainerIngressInstance_Basic (17.42s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes	21.249s
```

### CODE REVIEW

File: ibm/service/kubernetes/resource_ibm_container_ingress_secret_tls.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	d.Set("cert_crn", ingressSecretConfig.CRN)
+	d.Set("persistence", ingressSecretConfig.Persistence)
+	d.Set("domain_name", ingressSecretConfig.Domain)
+	d.Set("expires_on", ingressSecretConfig.ExpiresOn)
+	d.Set("status", ingressSecretConfig.Status)
+	d.Set("type", ingressSecretConfig.Type)
+	d.Set("user_managed", ingressSecretConfig.UserManaged)
+	d.Set("last_updated_timestamp", ingressSecretConfig.LastUpdatedTimestamp)
+
+	d.SetId(fmt.Sprintf("%s/%s/%s", cluster, secretName, secretNamespace))
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_ingress_secret_tls.go:179]: it is already set, no need to set again

File: ibm/service/kubernetes/resource_ibm_container_ingress_secret_tls.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+		Name:      secretName,
+		Namespace: secretNamespace,
+	}
+
+	err = ingressAPI.DeleteIngressSecret(params)
+	if err != nil {
+		return err
+	}
+
+	d.SetId("")
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_ingress_secret_tls.go:210]: no need to clear the id

File: ibm/service/kubernetes/resource_ibm_container_ingress_secret_opaque.go [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+			if fieldName, ok := opaqueFieldMap["field_name"]; ok {
+				fieldToAdd.Name = fieldName.(string)
+			}
+
+			if crn, ok := opaqueFieldMap["crn"]; ok {
+				fieldToAdd.CRN = crn.(string)
+			}
+
+			fieldsToAdd = append(fieldsToAdd, fieldToAdd)
```
[the remark this answers is not in the document]
Comment (theodoracheng) [reply ibm/service/kubernetes/resource_ibm_container_ingress_secret_opaque.go:165]: only the crn is required, and i set the schema for field.crn to required=true. i assume that means that terraform will do that validation then?

**Review by z0za — APPROVED** [review]

File: website/docs/r/container_ingress_secret_opaque.html.markdown [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+Registers an IBM Cloud Secrets Manager secret type certificate with your IBM Cloud Kubernetes Service or Red Hat OpenShift on IBM Cloud cluster. For more information about how opaque secrets can be used see [about Secrets Manager secrets](https://cloud.ibm.com/docs/containers?topic=containers-secrets#non-tls)
+
+## Example usage
+
+```terraform
+resource "ibm_container_ingress_secret_opaque" "secret" {
+  cluster="exampleClusterName"
+  secret_name="mySecretName"
+  secret_namespace="mySecretNamespace"
+  cert_crn="cert:region:crn"
```
Comment (hkantare) [inline website/docs/r/container_ingress_secret_opaque.html.markdown:19]: cert_crn is required /optional argument. its not defined in below section 
`Argument reference`

File: website/docs/r/container_ingress_secret_opaque.html.markdown [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  - `prefix` - (String) Prefix field name with Secrets Manager secret name
+
+## Attribute reference
+In addition to all argument reference list, you can access the following attribute reference after your resource is created.
+
+- `secret_type` - (String) The type of Kubernetes secret (Opaque).
+- `status` - (String) The Status of the secret.
+- `user_managed` - (Bool) Indicates whether the secret was created by a user.
+- `persistence`  - (Bool) Persist the secret data in your cluster. If the secret is later deleted from the command line or OpenShift web console, the secret is automatically re-created in your cluster.
+ `fields` - (Required, string) The fields of the opaque secret.
```
Comment (hkantare) [inline website/docs/r/container_ingress_secret_opaque.html.markdown:41]: In Attribute reference no need to mention "Required" and the field is not string type its a list.

Comment (hkantare) [reply website/docs/r/container_ingress_secret_opaque.html.markdown:41]: define something like this https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/is_vpc#security_group

File: website/docs/r/container_ingress_secret_opaque.html.markdown [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+}
+```
+
+## Argument reference
+Review the argument references that you can specify for your resource. 
+
+- `cluster` - (Required, String) The cluster ID.
+- `secret_name` - (Required, String) The name of the kubernetes secret.
+- `secret_namespace` - (Required, string) The namespace of the kubernetes secret.
+ `fields` - (Required, string) The fields of the opaque secret.
```
Comment (hkantare) [inline website/docs/r/container_ingress_secret_opaque.html.markdown:29]: Same as below comment its not "string" type

File: website/docs/r/container_ingress_secret_opaque.html.markdown [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  cert_crn="cert:region:crn"
+}
+```
+
+## Argument reference
+Review the argument references that you can specify for your resource. 
+
+- `cluster` - (Required, String) The cluster ID.
+- `secret_name` - (Required, String) The name of the kubernetes secret.
+- `secret_namespace` - (Required, string) The namespace of the kubernetes secret.
```
Comment (hkantare) [inline website/docs/r/container_ingress_secret_opaque.html.markdown:28]: persistence is defined in schema but missing in docs its optional argument

**Comment by hkantare** [comment]
General comment 

Can you move new res/ds to use the context 
ctx context.Context refer data_source_ibm_container_dedicated_host_flavor.go/resource_ibm_container_dedicated_host_pool for reference

**Pull request merged**